engine/inspect/inspect.go

```diff
@@ -0,0 +1,172 @@
+package inspect
+
+import (
+	"fmt"
+
+	"github.com/bmeg/grip/gripql"
+	"github.com/bmeg/grip/jsonpath"
+	"github.com/bmeg/grip/protoutil"
+	log "github.com/sirupsen/logrus"
+)
+
+func arrayEq(a, b []string) bool {
+	if len(a) != len(b) {
+		return false
+	}
+	for i := range a {
+		if a[i] != b[i] {
+			return false
+		}
+	}
+	return true
+}
+
+func contains(a []string, n string) bool {
+	for _, c := range a {
+		if c == n {
+			return true
+		}
+	}
+	return false
+}
+
+//PipelineSteps create an array, the same length at stmts that labels the
+//step id for each of the GraphStatements
+func PipelineSteps(stmts []*gripql.GraphStatement) []string {
+	out := []string{}
+	curState := 0
+	for _, gs := range stmts {
+		switch gs.GetStatement().(type) {
+		//These commands all change the position of the traveler. When that happens,
+		//we go to the next 'step' of the traversal
+		case *gripql.GraphStatement_V, *gripql.GraphStatement_E, *gripql.GraphStatement_Out,
+			*gripql.GraphStatement_In, *gripql.GraphStatement_OutE, *gripql.GraphStatement_InE,
+			*gripql.GraphStatement_Both, *gripql.GraphStatement_BothE, *gripql.GraphStatement_Select:
+			curState++
+		case *gripql.GraphStatement_Limit, *gripql.GraphStatement_As, *gripql.GraphStatement_Has,
+			*gripql.GraphStatement_HasId, *gripql.GraphStatement_HasKey, *gripql.GraphStatement_HasLabel,
+			*gripql.GraphStatement_Count, *gripql.GraphStatement_Skip, *gripql.GraphStatement_Distinct,
+			*gripql.GraphStatement_Range, *gripql.GraphStatement_Aggregate, *gripql.GraphStatement_Render,
+			*gripql.GraphStatement_Fields:
+		case *gripql.GraphStatement_LookupVertsIndex:
+		default:
+			log.Printf("Unknown Graph Statement: %T", gs.GetStatement())
+		}
+		out = append(out, fmt.Sprintf("%d", curState))
+	}
+	return out
+}
+
+//PipelineSteps identify the variable names each step can be aliasesed using
+//the as_ operation
+func PipelineAsSteps(stmts []*gripql.GraphStatement) map[string]string {
+	out := map[string]string{}
+	steps := PipelineSteps(stmts)
+
+	for i, gs := range stmts {
+		switch stmt := gs.GetStatement().(type) {
+		case *gripql.GraphStatement_As:
+			out[stmt.As] = steps[i]
+		}
+	}
+	return out
+}
+
+//PipelineStepOutputs identify the required outputs for each step in the traversal
+func PipelineStepOutputs(stmts []*gripql.GraphStatement) map[string][]string {
+
+	steps := PipelineSteps(stmts)
+	asMap := PipelineAsSteps(stmts)
+	onLast := true
+	out := map[string][]string{}
+	for i := len(stmts) - 1; i >= 0; i-- {
+		gs := stmts[i]
+		if onLast {
```

Comment:
embed this logic within individual cases in the switch statement.

Reply:
Added